backend/internal/userschema/service/userschemaservice.go

```diff
@@ -488,3 +492,394 @@ func logAndReturnServerError(
 	logger.Error(message, log.Error(err))
 	return &constants.ErrorInternalServerError
 }
+
+// ValidateUser validates user attributes against the user schema for the given user type.
+func (us *UserSchemaService) ValidateUser(userType string, userAttributes json.RawMessage) *serviceerror.ServiceError {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, userSchemaLoggerComponentName))
+
+	if userType == "" {
+		logger.Debug("User type is empty, skipping schema validation")
+		return nil
+	}
+
+	if userAttributes == nil {
+		logger.Debug("User has no attributes to validate")
+		return nil
+	}
+
+	if len(userAttributes) == 0 {
+		logger.Debug("User has no attributes to validate")
+		return nil
+	}
+
+	userSchema, err := store.GetUserSchemaByName(userType)
+	if err != nil {
+		if errors.Is(err, constants.ErrUserSchemaNotFound) {
+			logger.Debug("No schema found for user type, skipping validation", log.String("userType", userType))
+			return nil // Allow users without schema
+		}
+		return logAndReturnServerError(logger, "Failed to get user schema", err)
+	}
+
+	if err := validateUserAttributesAgainstSchema(userAttributes, userSchema.Schema); err != nil {
+		logger.Debug("Schema validation failed", log.Error(err), log.String("userType", userType))
+		return &constants.ErrorUserValidationFailed
+	}
+
+	logger.Debug("Schema validation successful", log.String("userType", userType))
+	return nil
+}
+
+// ValidateUserUniqueness validates the uniqueness constraints of user attributes.
+func (u *UserSchemaService) ValidateUserUniqueness(
+	userType string,
+	userAttributes json.RawMessage,
+	identifyUser func(map[string]interface{}) (*string, error),
+) (bool, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, userSchemaLoggerComponentName))
+
+	if userType == "" {
+		return true, nil
+	}
+
+	if userAttributes == nil {
+		return true, nil
+	}
+
+	if len(userAttributes) == 0 {
```

Comment:
len(userAttributes) should be already checking for nil. Hence nil check should be redundant

Reply:
Fixed with https://github.com/asgardeo/thunder/compare/50d21a542f55a15f2b3d369a94a7278220574656..8f5b6f50cfe613aad220dd8c60ccb836ffc6daba